pyteal/ast/opup.py

```diff
@@ -133,13 +138,21 @@ def ensure_budget(self, required_budget: Expr) -> Expr:
         return Seq(
             buffered_budget.store(required_budget + buffer),
             While(buffered_budget.load() > Global.opcode_budget()).Do(
-                self._construct_itxn()
+                self._construct_itxn(inner_fee=inner_fee)
             ),
         )
 
-    def maximize_budget(self, fee: Expr) -> Expr:
+    def maximize_budget(self, fee: Expr, inner_fee: Expr = Int(0)) -> Expr:
```

Comment:
@cusma + reviewers:  I see a backwards _incompatible_ change here and I'm opening the thread to solicit feedback on how to address.

* I _think_  the PR is backwards _incompatible_ due to its default behavior change:  inner transaction fee deducted from caller's account rather than app's account.
* If we agree the change _is_ backwards incompatible, then I think we must rework the PR.  I'm open to ideas and I can elaborate as is helpful.  Here's a few brainstormed thoughts (my preference **bolded**):
  * Add uniquely named methods (e.g. `maximimize_budget_with_caller_account`).  Implies deprecating the existing methods in favor of new, uniquely named methods.
  * **Define a class structure representing possible fee coverage scenarios.**  Modify existing methods to default to the type that matches existing behavior.
* Regardless of implementation approach taken, I think we need to agree on the cases to support.  Here's the cases I see:
  * Paid by caller:  `inner_fee` = 0.
  * Paid by app until app account exhausted:  `inner_fee` _not_ set (as is behavior).
  * Paid by app until max fee reached:  `inner_fee` set with value > 0.

Reply:
@michaeldiamant I preferred to have `inner_fee=Int(0)` as default because this automatically preserves App's funds, differently from what happens with the _as-is` behaviour, that could potentially expose the App to unwanted fee draining.

I see your point, let me elaborate on that step by step:

- I agree we slightly change the default behaviour but I'm not sure if this _literally breaks_ the API backward compatibility.
- If we think this _is_ backward incompatible instead, then there could be another option: replace `inner_fee=Int(0)` with `inner_fee=Global.min_txn_fee()`: in this case we preserve the exact same existing behaviour but we loose the "App fund usage" as default.
- On your last point:
    - `inner_fee` = 0: I agree this implies all Inner Transactions fee are paid by the caller, either for `OpUp.ensure_budget` or `OpUp.maximize_budget`;
    - `inner_fee` not set: this implies Inner Transactions fee are "paid by app until app account exhausted" _just_ for `OpUp.ensure_budget`, since `OpUp.maximize_budget` always requires a `fee` parameter (again, not very straightforward naming) to cap the fee consumptions. The point is that we still have to decide "how much" the `OpUp` operations cost: I suppose the answer is `min_txn_fee` as default.
    - `inner_fee` > 0: this implies Inner Transactions fee are "paid by app until max fee reached" _just_ for `OpUp.maximize_budget`.

Now my question is: for the `OpUp` operations, should we enable Inner Transactions fee to be paid by the caller _just_ for the `OpUp.maximize_budget` option? Under `OpUp.ensure_budget` the Inner Transaction fee could be different depending on the execution path, while with `OpUp.maximize_budget` the caller would always know in advance how much fee must be paid with the App call. In the `OpUp.ensure_budget` case, due to execution path dependency, it could be difficult (although possible) for a client to correctly set a sufficient App call fee.

Reply:
@cusma I'm responding inline below.  For now, I'm putting aside implementation concerns to see if we can agree on prerequisite topics.

>  I'm not sure if this literally breaks the API backward compatibility.

* I'd like to agree with you, but the proposal silently modifies the default behavior for an already released API.
* I understand and agree with why the PR opts for a different default.  Unfortunately, doing so is a breaking change.
* _Possible_ middle ground provided reviewers agree:  Deprecate existing methods in favor of new methods with a safer default or all required parameters.   

> replace inner_fee=Int(0) with inner_fee=Global.min_txn_fee(): in this case we preserve the exact same existing behaviour

Unfortunately, the proposal does _not_ preserve existing behavior when the caller overpays.

> Under OpUp.ensure_budget the Inner Transaction fee could be different depending

Does the following rationale convince you that both OpUp methods ought to support payor configuration?  If not, can you add more context to help me see the intuition?
* I agree with the quoted statement though I think it's orthogonal to (what I believe to be) the central question:  How can a smart contract writer configure the payor configuration appropriate for his/her smart contract?
* Ultimately, _some_ account(s) must pay the fees.  Since `OpUp` is a general purpose mechanism supporting, I think it's appropriate for all methods issuing inner-transactions to support payor configuration.

Reply:
@michaeldiamant my answers:

> Possible middle ground provided reviewers agree: Deprecate existing methods in favor of new methods with a safer default or all required parameters.

Yes, this could be a way. Once we define the requirements we can keep this option in mind.

> Unfortunately, the proposal does not preserve existing behavior when the caller overpays.

Right, I neglected this corner case due to my bias towards "prioritizing caller payment should be preferable", but you are right in saying that this is not the _exact_ same behavior. So we can discard the `inner_fee=Global.min_txn_fee()` option.

> Ultimately, some account(s) must pay the fees. Since `OpUp` is a general purpose mechanism supporting, I think it's appropriate for all methods issuing inner-transactions to support payor configuration.

Agree, I was maybe overthinking on the "client side", but staying strictly on the "smart contract writer" one I would say that would be preferable to support inner-transaction payor configuration in any case.

Reply:
@michaeldiamant thanks!

Due to a new little member of the family 🐣🍼 I have to wind down my commitment on keeping this PR going forward, but I'll try to  follow the discussion and implementation progress every now and then.

Reply:
I just want to chime in and say my personal preference for avoiding a breaking change is to make the `inner_fee` parameter default to `None`, which would indicate `TxnField.fee` should not be set on the inner txn (the same as the current behavior in master). Only by passing in a non-`None` value would the user receive code that explicitly sets the fee.

Reply:
I implemented the None default param for this in https://github.com/algorand/pyteal/pull/566 but since the pr is on the `cusma` repo I don't think I can target it easily.